frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java

```diff
@@ -110,27 +130,67 @@ public void onSuccess() {
     });
   }
 
+  @BackgroundableQueuedElsewhere // passed on to `executeAfterRunningPrePushHandlers`
+  @UIEffect
+  public void pushAndQuit() {
+
+    String title = getNonHtmlString("string.GitMachete.GitPushDialog.task-title");
+    executeAfterRunningPrePushHandlers(new SideEffectingBackgroundable(myProject, title, "push") {
+      @Override
+      @UIThreadUnsafe
+      public void doRun(ProgressIndicator indicator) {
+        myController.push(isForcePushRequired);
+      }
+    });
+  }
+
   @Override
   public @Nullable VcsPushOptionValue getAdditionalOptionValue(PushSupport support) {
     return null;
   }
 
-  private String getPushActionName() {
-    return isForcePushRequired ? "Force _Push" : "_Push";
+  private String getPushActionName(boolean isAndQuit) {
+    if (isAndQuit) {
+      return (isForcePushRequired ? "Force Push" : "Push") + " _and Quit";
+    } else {
+      return isForcePushRequired ? "Force _Push" : "_Push";
+    }
   }
 
   private class PushSwingAction extends AbstractAction {
 
+    private final Runnable pushRunnable;
+
+    @UIEffect
+    PushSwingAction(boolean isAndQuit) {
```

Comment:
Maybe `hasAndQuit`?

Reply:
Actually... maybe create 2 separate classes? `PushSwingAction` and `PushAndQuitSwingAction`